tensorflow/stream_executor/rocm/rocm_blas.h

```diff
@@ -97,6 +98,20 @@ class ROCMBlas : public blas::BlasSupport {
                               /*err_on_failure=*/false, args...);
   }
 
+  // A helper struct to for type conversion in DoBlasGemmBatchedInternal
+  template <typename T>
+  struct EigenHalfToRocBlasHalf{
+    using type = T;
+  };
+  
+  // A helper allocation funciton to convert raw pointers memory layout to strided flavor 
+  template <typename T>
+  port::Status AllocateStridedBuffer(
+      const std::vector<typename EigenHalfToRocBlasHalf<T>::type *>& raw_ptrs, int batch_count, 
+      uint64_t batch_stride, ScratchAllocator* scrach_allocator, Stream * stream, 
+      std::unique_ptr<TemporaryDeviceMemory<typename EigenHalfToRocBlasHalf<T>::type>>& temp_memory,
+      DeviceMemory<typename EigenHalfToRocBlasHalf<T>::type>& device_memory);
```

Comment:
@jerryyin I don't think non-const references would be accepted upstream. `temp_memory` and `device_memory` must be changed.

Reply:
also this helper function deserves better documentation on the semantic of each parameter. it seems it may do non-trivial memory allocation/deallocation.

Reply:
@whchung This is a helper function used in DoBlasGemmBatchedInternal with very narrowed scope. Both `device_memory` and `temp_memory` are assigned in the function and serve as return values. They won't be able to hold return value if I declare them as const. How would you suggest I'd change here?

Reply:
`device_memory` shall be passed in as a pointer.

and reviewing the logic, do we really need `temp_memory`? it seems we can remove this parameter and solely rely on `ScratchAllocator`?

Reply:
For `device_memory`, I can pass it in as a pointer. Still curious at why it needs to be a pointer instead of a reference? To me, passing as reference is syntax-convenient, and save us trouble for callee to do the nullptr check. 

`temp_memory` is used only when scratch_allocator is a nullptr. Since scratch_allocator is passed all way down from DoBlasGemmBatched() as a pointer, I didn't find a guarantee that it couldn't be a nullptr. When scratch_allocator is actually a nullptr, temp_memory will be the owner of the gpu memory. The only reason it needs to be passed in instead of locally allocated is due to the unique pointer manages its life cycle. When the temporary memory is allocated outside, it gives guarantee that the memory is still available by the end of DoBlasGemmBatchedInternal(), ie, the memory is valid when we make the actual gemm call into rocBlas. I learned this from reading `cuda_blas.cc`. Let me know if you think there is a neater way to do the temporary memory allocation here.

Reply:
non-const reference is simply not accepted upstream so we'll have to abide this rule.

Reply:
how about we change the interface so it returns `port::StatusOr<std::unique_ptr<DeviceMemory>>`? in that way we can pass `temp_memory` in by value, removing non-const reference.

Reply:
Thanks for explaining. Now that I understand it is a policy, it looks like either returning status or return a device memory will make the other pass by pointer. Both way works for me. For the perspective of minimal change of the current code base, I'll probably just pass the memory by pointer and return the (cheaper) status by value.